File: backend/app/models/pair_trade.py
from datetime import datetime, timedelta, timezone
from enum import Enum
from typing import Optional, Dict, Any, List
from pydantic import BaseModel, Field, PositiveFloat

from app.utils.time_utils import get_utc_now


# 定義 UTC+8 時區 (僅用於向後兼容)
UTC_PLUS_8 = timezone(timedelta(hours=8))


# 獲取 UTC+8 當前時間的函數 - 已移至 time_utils.py
# def get_utc_plus_8_now():
#     """獲取 UTC+8 時區的當前時間"""
#     return datetime.now(UTC_PLUS_8)


class TradeStatus(str, Enum):
    """交易狀態枚舉"""
    PENDING = "pending"  # 等待執行
    ACTIVE = "active"    # 活躍中
    CLOSED = "closed"    # 已平倉
    FAILED = "failed"    # 執行失敗


class TradePosition(BaseModel):
    """交易持倉信息"""
    symbol: str  # 交易對符號，例如 BTCUSDT
    side: str    # 方向，BUY 或 SELL
    quantity: float  # 數量
    entry_price: float  # 入場價格
    current_price: float = 0  # 當前價格
    exit_price: float = 0  # 平倉價格
    pnl: float = 0  # 盈虧金額
    pnl_percent: float = 0  # 盈虧百分比
    entry_order_id: str  # 開倉訂單ID
    notional_value: float  # 名義價值（數量 * 價格）
    entry_fee: float = 0  # 開倉手續費
    exit_fee: float = 0  # 平倉手續費
    exit_order_id: str = ""  # 平倉訂單ID
    leverage: float = 1  # 槓桿倍數


class PairTrade(BaseModel):
    """配對交易模型"""
    id: Optional[str] = None  # 交易ID（用於API響應）
    user_id: str  # 用戶ID
    name: str  # 交易名稱
    status: TradeStatus = TradeStatus.PENDING  # 交易狀態
    max_loss: float  # 最大虧損額度（USDT）
    stop_loss: float  # 止損百分比
    take_profit: float  # 止盈百分比
    # 新增停利相關欄位
    trailing_stop_enabled: bool = False  # 是否啟用停利保護
    trailing_stop_level: float = 0  # 停利水位百分比
    long_position: Optional[TradePosition] = None  # 多單持倉
    short_position: Optional[TradePosition] = None  # 空單持倉
    total_pnl_value: float = 0  # 總盈虧金額
    total_ratio_percent: float = 0  # 總盈虧百分比
    total_fee: float = 0  # 總手續費
    total_entry_fee: float = 0  # 開倉總手續費
    total_exit_fee: float = 0  # 平倉總手續費
    max_ratio: float = 0  # 最高比率（多空價格比）
    min_ratio: float = 0  # 最低比率（多空價格比）
    mae: float = 0  # 最大不利變動 (MAE)
    mfe: float = 0  # 最大有利變動 (MFE)
    net_pnl: float = 0  # 淨盈虧（扣除手續費）
    total_pnl: float = 0  # 總盈虧（未扣除手續費）
    leverage: float = 1  # 槓桿倍數
    long_leverage: float = 1  # 多單槓桿倍數
    short_leverage: float = 1  # 空單槓桿倍數
    risk_reward_ratio: float = 0  # 風險收益比
    net_risk_reward_ratio: float = 0  # 淨風險收益比
    # 使用 UTC 時間
    created_at: datetime = Field(
        default_factory=get_utc_now
    )  # 創建時間
    updated_at: datetime = Field(
        default_factory=get_utc_now
    )  # 更新時間
    closed_at: Optional[datetime] = None  # 平倉時間
    close_reason: Optional[str] = None  # 平倉原因


class PairTradeCreate(BaseModel):
    """創建配對交易的請求模型"""
    name: Optional[str] = ""  # 交易名稱（可選，如果為空則自動生成）
    max_loss: float  # 最大虧損額度（USDT），1R unit
    stop_loss: float  # 止損百分比
    take_profit: float  # 止盈百分比
    long_symbol: str  # 多單交易對符號
    short_symbol: str  # 空單交易對符號
    long_leverage: float = 1  # 多單槓桿倍數
    short_leverage: float = 1  # 空單槓桿倍數
    margin_type: str = "ISOLATED"  # 保證金類型，ISOLATED 或 CROSSED
    test_mode: bool = False  # 測試模式，不執行實際交易


class PairTradeResponse(BaseModel):
    """配對交易響應模型"""
    id: str
    name: str
    status: str
    max_loss: float
    stop_loss: float
    take_profit: float
    # 新增停利相關欄位
    trailing_stop_enabled: bool = False
    trailing_stop_level: float = 0
    long_position: Optional[Dict[str, Any]] = None
    short_position: Optional[Dict[str, Any]] = None
    total_pnl_value: float
    total_ratio_percent: float
    total_fee: float = 0  # 總手續費
    total_entry_fee: float = 0  # 開倉總手續費
    total_exit_fee: float = 0  # 平倉總手續費
    max_ratio: float = 0  # 最高比率（多空價格比）
    min_ratio: float = 0  # 最低比率（多空價格比）
    mae: float = 0  # 最大不利變動 (MAE)
    mfe: float = 0  # 最大有利變動 (MFE)
    net_pnl: float = 0  # 淨盈虧（扣除手續費）
    total_pnl: float = 0  # 總盈虧（未扣除手續費）
    leverage: float = 1  # 槓桿倍數
    risk_reward_ratio: float = 0  # 風險收益比
    net_risk_reward_ratio: float = 0  # 淨風險收益比
    created_at: datetime
    updated_at: datetime
    closed_at: Optional[datetime] = None
    close_reason: Optional[str] = None


class PairTradeSummaryResponse(BaseModel):
    total_active: int
    total_closed: int
    overall_pnl: float
    # 可在此添加更多匯總信息


class PairTradeSettingsUpdate(BaseModel):
    """
    用於更新配對交易止盈/止損設定的模型
    """
    take_profit: Optional[PositiveFloat] = Field(None, description="新的止盈百分比 (例如輸入 5 表示 5%)")
    stop_loss: Optional[PositiveFloat] = Field(None, description="新的止損百分比 (例如輸入 3 表示 3%)")
    # 新增停利相關欄位
    trailing_stop_enabled: Optional[bool] = Field(None, description="是否啟用停利保護")
    trailing_stop_level: Optional[float] = Field(None, ge=0, description="停利水位百分比 (例如輸入 8 表示 8%)")

    class Config:
        extra = "forbid"


class BulkActionRequest(BaseModel):
    trade_ids: List[str]
    action: str  # e.g., "close", "delete"

File: frontend/src/lib/api/pair-trade.ts
import apiClient from './client';

// 配對交易創建請求接口
export interface PairTradeCreateRequest {
  name: string;
  max_loss: number;
  stop_loss: number;
  take_profit: number;
  long_symbol: string;
  short_symbol: string;
  leverage?: number;
  long_leverage?: number;
  short_leverage?: number;
  margin_type?: string;
}

// 交易持倉接口
export interface TradePosition {
  symbol: string;
  side: string;
  quantity: number;
  entry_price: number;
  current_price: number;
  exit_price: number;
  pnl: number;
  pnl_percent: number;
  entry_order_id: string;
  notional_value: number;
  entry_fee: number;
  exit_fee: number;
  exit_order_id: string;
  leverage: number;
}

// 配對交易接口
export interface PairTrade {
  id: string;
  name: string;
  status: string;
  max_loss: number;
  stop_loss: number;
  take_profit: number;
  trailing_stop_enabled: boolean;
  trailing_stop_level: number;
  long_position: TradePosition | null;
  short_position: TradePosition | null;
  total_pnl_value: number;
  total_ratio_percent: number;
  total_fee: number;  // 總手續費
  total_entry_fee: number;  // 開倉總手續費
  total_exit_fee: number;   // 平倉總手續費
  max_ratio: number;  // 最高比率（多空價格比）
  min_ratio: number;  // 最低比率（多空價格比）
  mae: number;  // 最大不利變動 (MAE)
  mfe: number;  // 最大有利變動 (MFE)
  created_at: string;
  updated_at: string;
  closed_at: string | null;
  close_reason: string | null;
}

export interface CreatePairTradeParams {
  name?: string;
  max_loss: number;
  stop_loss: number;
  take_profit: number;
  long_symbol: string;
  short_symbol: string;
  leverage?: number;
  long_leverage?: number;
  short_leverage?: number;
  margin_type?: string;
  test_mode?: boolean;
}

/**
 * 創建配對交易
 * @param params 配對交易參數
 * @returns 創建的配對交易
 */
export const createPairTrade = async (params: CreatePairTradeParams): Promise<PairTrade> => {
  try {
    const response = await apiClient.post<PairTrade>('/pair-trades', params);
    return response.data;
  } catch (error) {
    console.error('創建配對交易失敗:', error);
    throw error;
  }
};

/**
 * 獲取所有配對交易
 * @param status 可選的交易狀態過濾條件
 * @returns 配對交易列表
 */
export const getPairTrades = async (status?: string): Promise<PairTrade[]> => {
  try {
    const url = status ? `/pair-trades?status=${status}` : '/pair-trades';
    const response = await apiClient.get<PairTrade[]>(url);
    return response.data;
  } catch (error) {
    console.error('獲取配對交易列表失敗:', error);
    throw error;
  }
};

/**
 * 獲取指定的配對交易
 * @param id 配對交易ID
 * @returns 配對交易
 */
export const getPairTrade = async (id: string): Promise<PairTrade> => {
  try {
    const response = await apiClient.get<PairTrade>(`/pair-trades/${id}`);
    return response.data;
  } catch (error) {
    console.error('獲取配對交易失敗:', error);
    throw error;
  }
};

/**
 * 更新配對交易
 * @param id 配對交易ID
 * @returns 更新後的配對交易
 */
export const updatePairTrade = async (id: string): Promise<PairTrade> => {
  try {
    const response = await apiClient.put<PairTrade>(`/pair-trades/${id}`);
    return response.data;
  } catch (error) {
    console.error('更新配對交易失敗:', error);
    throw error;
  }
};

/**
 * 平倉配對交易
 * @param id 配對交易ID
 * @returns 平倉後的配對交易
 */
export const closePairTrade = async (id: string): Promise<PairTrade> => {
  try {
    const response = await apiClient.delete<PairTrade>(`/pair-trades/${id}`);
    return response.data;
  } catch (error) {
    console.error('平倉配對交易失敗:', error);
    throw error;
  }
};

// 定義更新交易設定的請求參數接口
export interface UpdatePairTradeSettingsParams {
  take_profit?: number;
  stop_loss?: number;
  trailing_stop_enabled?: boolean;
  trailing_stop_level?: number;
}

/**
 * 更新配對交易的止盈/止損設定
 * @param tradeId 配對交易ID
 * @param settings 包含新的止盈/止損值的物件
 * @returns 更新後的配對交易
 */
export const updatePairTradeSettings = async (
  tradeId: string, 
  settings: UpdatePairTradeSettingsParams
): Promise<PairTrade> => {
  try {
    const response = await apiClient.put<PairTrade>(
      `/pair-trades/${tradeId}/settings`,
      settings
    );
    return response.data;
  } catch (error) {
    console.error(`更新交易 ${tradeId} 設定失敗:`, error);
    throw error;
  }
};